partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt

```diff
@@ -681,34 +686,75 @@ internal class PlanTyper(
         }
 
         override fun visitRexOpCase(node: Rex.Op.Case, ctx: StaticType?): Rex {
-            // Type branches and prune branches known to never execute
-            val newBranches = node.branches.map { visitRexOpCaseBranch(it, it.rex.type) }
-                .filterNot { isLiteralBool(it.condition, false) }
+            // Rewrite CASE-WHEN branches
+            val oldBranches = node.branches.toTypedArray()
+            var newBranches = mutableListOf<Rex.Op.Case.Branch>()
+            val typer = DynamicTyper()
+            for (i in oldBranches.indices) {
+
+                // Type the branch
+                var branch = oldBranches[i]
+                branch = visitRexOpCaseBranch(branch, branch.rex.type)
+
+                // Check if branch condition is a literal
+                val condition = branch.condition
+                when (boolOrNull(condition.op)) {
+                    true -> return branch.rex // fold
```

Comment:
I think there's a bug here. Let's say a CASE has three branches (the first two being booleans (but not constants) and the third being a boolean of literal true). With this logic, it seems like we'd accumulate the first two visited branches, but for the third, we'd immediately return it's result.

Reply:
I'm going to leave constant folding `true` out of this considering the urgency